File: go/roman-numerals/roman_numerals.go
/*
Package romannumerals provides a method to convert a number to roman nuermals.
*/
package romannumerals

import (
	"errors"
	"fmt"
)

var romanSet = [][]string{
	[]string{"I", "V", "X"},
	[]string{"X", "L", "C"},
	[]string{"C", "D", "M"},
	[]string{"M", "", ""},
}

// ToRomanNumeral converts a numeral to roman.
func ToRomanNumeral(num int) (string, error) {
	if 0 < num && num <= 3000 {
		output := ""
		for pos := 1; num > 0; pos++ {
			digit := num % 10
			num = num / 10
			output = romanDigit(digit, pos) + output
		}
		return output, nil
	}
	return "", errors.New("invalid number")
}

// romanDigit converts a single digit to roman number based on its position
func romanDigit(num int, pos int) string {
	set := romanSet[pos-1]
	switch num {
	case 1:
		return fmt.Sprintf("%s", set[0])
	case 2:
		return fmt.Sprintf("%s%s", set[0], set[0])
	case 3:
		return fmt.Sprintf("%s%s%s", set[0], set[0], set[0])
	case 4:
		return fmt.Sprintf("%s%s", set[0], set[1])
	case 5:
		return fmt.Sprintf("%s", set[1])
	case 6:
		return fmt.Sprintf("%s%s", set[1], set[0])
	case 7:
		return fmt.Sprintf("%s%s%s", set[1], set[0], set[0])
	case 8:
		return fmt.Sprintf("%s%s%s%s", set[1], set[0], set[0], set[0])
	case 9:
		return fmt.Sprintf("%s%s", set[0], set[2])
	}
	return ""
}

File: elixir/roman-numerals/lib/roman_numerals.ex
defmodule RomanNumerals do
  @doc """
  Convert the number to a roman number.
  """
  @roman_numerals [
    %{1 => 'I', 5 => 'V', 10 => 'X'},
    %{1 => 'X', 5 => 'L', 10 => 'C'},
    %{1 => 'C', 5 => 'D', 10 => 'M'},
    %{1 => 'M'}
  ]

  @spec numeral(pos_integer) :: String.t()
  def numeral(number) do
    numeral_to_roman(number, @roman_numerals)
    |> to_string()
  end

  defp numeral_to_roman(number, roman_set) when number > 0 do
    numeral_to_roman(div(number, 10), tl(roman_set)) ++ romanize(rem(number, 10), hd(roman_set))
  end

  defp numeral_to_roman(_, _) do
    ''
  end

  defp romanize(number, roman_set) do
    case number do
      0 -> ''
      1 -> roman_set[1]
      2 -> roman_set[1] ++ roman_set[1]
      3 -> roman_set[1] ++ roman_set[1] ++ roman_set[1]
      4 -> roman_set[1] ++ roman_set[5]
      5 -> roman_set[5]
      6 -> roman_set[5] ++ roman_set[1]
      7 -> roman_set[5] ++ roman_set[1] ++ roman_set[1]
      8 -> roman_set[5] ++ roman_set[1] ++ roman_set[1] ++ roman_set[1]
      9 -> roman_set[1] ++ roman_set[10]
    end
  end
end
